core/metacling/src/TCling.cxx

```diff
@@ -1122,7 +1122,59 @@ static GlobalModuleIndex *loadGlobalModuleIndex(SourceLocation TriggerLoc, cling
          RecreateIndex = true;
       }
       if (RecreateIndex) {
-         GlobalModuleIndex::writeIndex(CI.getFileManager(), CI.getPCHContainerReader(), ModuleIndexPath);
+         cling::Interpreter::PushTransactionRAII deserRAII(&interp);
+         clang::GlobalModuleIndex::UserDefinedInterestingIDs IDs;
+
+         struct DefinitionFinder : public RecursiveASTVisitor<DefinitionFinder> {
+            DefinitionFinder(clang::GlobalModuleIndex::UserDefinedInterestingIDs& IDs,
+                             clang::TranslationUnitDecl* TU) : DefinitionIDs(IDs) {
+               TraverseDecl(TU);
+            }
+            bool VisitNamedDecl(NamedDecl *ND) {
+               for (auto R : ND->redecls()) {
+                  if (!R->isFromASTFile())
+                     continue;
+                  if (TagDecl *TD = llvm::dyn_cast<TagDecl>(R)) {
+                     if (TD->isCompleteDefinition())
+                        Register(TD);
+                  } else if (NamespaceDecl *NSD = llvm::dyn_cast<NamespaceDecl>(R))
+                     Register(NSD, /*AddSingleEntry=*/ false);
+                  else if (TypedefNameDecl *TND = dyn_cast<TypedefNameDecl>(R))
+                     Register(TND);
```

Comment:
Can't typedef decl be repeated (in several header files and thus several modules)?

Reply:
Yes, but it should define the same entity, otherwise modules ODR hash will complain. So whichever we pick should be fine. I suspect that may lead to loading the 'wrong' module to resolve the intended typedef but cannot do anything better.

Alternatively, we could load all modules which have this typedef name -- that would be redundant because if the typedefs are all the same what's the point; if the typedefs are different we will get an ODR violation diagnostic.

Reply:
Do we need to load the pcm for the target of the typedef?

Reply:
Yes, we need to load something that has a clang::Decl for the missing identifier